conda_lock/conda_lock.py

```diff
@@ -23,42 +31,130 @@
 DEFAULT_PLATFORMS = ["osx-64", "linux-64", "win-64"]
 
 
-def solve_specs_for_arch(channels: List[str], specs: List[str], platform: str) -> str:
+def ensure_conda(conda_executable: Optional[str]):
+    if conda_executable:
+        if pathlib.Path(conda_executable).exists():
+            return conda_executable
+
+    conda_executable = shutil.which("conda")
+    if conda_executable:
+        return conda_executable
+
+    conda_executable = shutil.which("conda.exe")
+    if conda_executable:
+        return conda_executable
+
+    logging.info(
+        "No existing conda installation found.  Installing the standalone conda solver"
+    )
+    return install_conda_exe()
+
+
+def install_conda_exe():
+    conda_exe_prefix = "https://repo.anaconda.com/pkgs/misc/conda-execs"
+    if platform.system() == "Linux":
+        conda_exe_file = "conda-latest-linux-64.exe"
+    elif platform.system() == "Darwin":
+        conda_exe_file = "conda-latest-osx-64.exe"
+    elif platform.system() == "NT":
+        conda_exe_file = "conda-latest-win-64.exe"
+    else:
+        # TODO: Support windows here
+        raise ValueError(f"Unsupported platform: {platform.system()}")
+
+    resp = requests.get(f"{conda_exe_prefix}/{conda_exe_file}", allow_redirects=True)
+    resp.raise_for_status()
+    target_filename = os.path.expanduser(pathlib.Path(__file__).parent / "conda.exe")
+    with open(target_filename, "wb") as fo:
+        fo.write(resp.content)
+    st = os.stat(target_filename)
+    os.chmod(target_filename, st.st_mode | stat.S_IXUSR)
+    return target_filename
+
+
+CONDA_PKGS_DIRS = None
+
+
+def conda_pkgs_dir():
+    global CONDA_PKGS_DIRS
+    if CONDA_PKGS_DIRS is None:
+        CONDA_PKGS_DIRS = tempfile.TemporaryDirectory()
+        atexit.register(CONDA_PKGS_DIRS.cleanup)
+        return CONDA_PKGS_DIRS
+
+
+def conda_env_override():
     env = dict(os.environ)
-    with tempfile.TemporaryDirectory() as CONDA_PKGS_DIRS:
-        env.update(
-            {
-                "CONDA_SUBDIR": platform,
-                "CONDA_PKGS_DIRS": CONDA_PKGS_DIRS,
-                "CONDA_UNSATISFIABLE_HINTS_CHECK_DEPTH": "0",
-                "CONDA_ADD_PIP_AS_PYTHON_DEPENDENCY": "False",
-            }
+    env.update(
+        {
+            "CONDA_SUBDIR": platform,
+            "CONDA_PKGS_DIRS": conda_pkgs_dir(),
+            "CONDA_UNSATISFIABLE_HINTS_CHECK_DEPTH": "0",
+            "CONDA_ADD_PIP_AS_PYTHON_DEPENDENCY": "False",
+        }
+    )
+    return env
+
+
+def solve_specs_for_arch(
+    conda: PathLike, channels: List[str], specs: List[str], platform: str
+) -> dict:
+    args: MutableSequence[PathLike] = [
+        conda,
+        "create",
+        "--prefix",
+        str(pathlib.Path(conda_pkgs_dir()).joinpath("prefix")),
```

Comment:
With the `PathLike` type we can remove the `str()` here, right?
(I added that and it is quite ugly, sorry.)

Reply:
```suggestion
        pathlib.Path(conda_pkgs_dir()).joinpath("prefix"),
```